### PR TITLE
[doc] Document list generation

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #709 #941 #986 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: docs/internal.rst [hunk 1]
```diff
@@ -1,12 +1,81 @@
...
+---------------------
+
+Struct-fors in Taichi loop over all active elements of a (sparse) data structure **in parallel**.
+Evenly distributing work onto processor cores is challenging on sparse data structures:
+naively splitting an irregular tree into pieces can easily lead to
+partitions with drastically different numbers of leaf elements.
+
+Our strategy is to generate lists of SNode elements layer by layer.
+This flattens the data structure leaf elements into a 1D dense array, circumventing the irregularity of
+incomplete trees. Then we can simply invoke a regular **parallel for** over the list.
```
[the remark this answers is not in the document]
Comment (k-ye) [reply docs/internal.rst:19]: It's on GPU, and these lists are updated in a kernel. There are multiple lists so I think plural form is the correct one. I'd suggest to say `lists of active SNode elements layer by layer`.

[the remark this answers is not in the document]
Comment (archibate) [reply docs/internal.rst:19]:
```suggestion
Our strategy is to generate several lists of active SNode elements on GPU, layer by layer.
This flattens the data structure leaf elements into a 1D dense array, circumventing the irregularity of
incomplete trees. Then we can simply invoke a regular **parallel for** over the list.
```

File: misc/listgen_demo.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+ti.init(print_ir=True)
+
+x = ti.var(ti.i32)
+ti.root.dense(ti.i, 4).bitmasked(ti.i, 4).place(x)
+
+@ti.kernel
+def func():
+    for i in x:
+        print(i)
```
[the remark this answers is not in the document]
Comment (yuanming-hu) [reply misc/listgen_demo.py:11]: That's true. We are just illustrating list generation here.

File: docs/internal.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuanming-hu) [reply docs/internal.rst:19]: I would stick to the original version since the generation can happen on either CPUs or GPUs, depending on `arch`.

**Comment by yuanming-hu** [comment]
Thanks for the reviews! I'm merging this, so that I can add more details on the internal designs (for #986) following this PR.

**Pull request merged**